src/main.cpp

```diff
@@ -17,19 +17,329 @@
+
+    void resumed() override
+    {
+        std::cerr << "resume" << std::endl;
+
+        if (auto active_session = weak_active_session().lock())
+            active_session->set_lifecycle_state(mir_lifecycle_state_resumed);
+    }
+
+    void started() override
+    {
+    }
+
+    void ready_for_user_input() override
+    {
+    }
+
+    void stop_receiving_input() override
+    {
+    }
+
+    std::weak_ptr<mir::scene::Session> weak_active_session()
+    {
+        return focus_controller->focused_session();
+    }
+
+    std::shared_ptr<mir::shell::FocusController> const focus_controller;
+};
+
+struct StubCookie : public mir::cookie::Cookie
+{
+    uint64_t timestamp() const override
+    {
+        return 0;
+    }
+
+    std::vector<uint8_t> serialize() const override
+    {
+        // FIXME Workaround until mir fixes its internal message bliting
+        // This size needs to equal mir/src/include/cookie/mir/cookie/blob.h::default_blob_size
+        return std::vector<uint8_t>(29);
+    }
+};
+
+struct StubCookieAuthority : public mir::cookie::Authority
+{
+    std::unique_ptr<mir::cookie::Cookie> make_cookie(uint64_t const& timestamp) override
+    {
+        return std::unique_ptr<StubCookie>(new StubCookie());
+    }
+
+    std::unique_ptr<mir::cookie::Cookie> make_cookie(std::vector<uint8_t> const& raw_cookie) override
+    {
+        return std::unique_ptr<StubCookie>(new StubCookie());
+    }
+};
 
 int main(int argc, char *argv[])
 try
 {
-    auto const config = std::make_shared<usc::Server>(argc, argv);
+    auto const config = std::make_shared<mir::Server>();
+
+    auto version_flag = pre_init(miral::CommandLineOption(
+        [](bool is_set)
+        {
+            if (is_set)
+            {
+                std::cerr << "unity-system-compositor " << USC_VERSION << std::endl;
+                exit(EXIT_SUCCESS);
+            }
+        },
+        "version", "Show version of Unity System Compositor"));
+
+    version_flag(*config);
+
+    miral::CommandLineOption blacklist_flag{
+        &usc::validate_video_driver,
+        "blacklist", "Video driver blacklist regex to use", ""
+    };
+    blacklist_flag(*config);
+
+    // socket file settings
+    std::string socket_file;
+    bool no_socket_file;
+    bool public_socket;
+
+    usc::ExtractSocketConfig extract_socket_config{
+        [&socket_file, &no_socket_file](std::string file_flag, bool no_file_flag)
+        {
+            socket_file = file_flag;
+            no_socket_file = no_file_flag;
+        }
+    };
+    extract_socket_config(*config);
+
+    miral::CommandLineOption public_socket_flag{
+        [&public_socket, no_socket_file](bool is_set)
+        {
+            public_socket = is_set && !no_socket_file;
+        },
+        "public-socket",
+        "Make the socket file publicly writable",
+        true
+    };
+    public_socket_flag(*config);
+
+    // spinner settings
+    std::string spinner_path;
+    miral::CommandLineOption spinner_flag{
```

Comment:
Should probably be `pre_init` just to be sure that it is available by the time it's used (may be worth checking this in any case, similar to what's been done for `--enable-hardware-cursor`).